File: apps/mobile/lib/features/profile/data/profile_service.dart
// Profile Service — VSP Mobile App
//
// API calls for golfer profile endpoints (/profiles/*).
// All methods return structured results or throw VspApiException.

import '../../../core/network/api_client.dart';
import 'profile_dto.dart';

/// Profile service — wraps /profiles/* API calls.
class ProfileService {
  final ApiClient _apiClient;

  ProfileService({required ApiClient apiClient}) : _apiClient = apiClient;

  // ─── Profile ─────────────────────────────────────────────────────────────────

  /// GET /profiles/me — fetch the authenticated golfer's profile.
  ///
  /// Backend auto-creates a default profile on first access.
  /// All distance values (e.g. driverDistance) are returned in METERS
  /// regardless of the golfer's display unit preference.
  Future<ProfileDTO> getProfile() async {
    final response = await _apiClient.get('/profiles/me');
    return ProfileDTO.fromJson(response as Map<String, dynamic>);
  }

  /// PUT /profiles/me — update the authenticated golfer's profile.
  ///
  /// Supports partial updates (only send fields you want to change).
  /// Idempotency key is required for offline-queued updates to ensure
  /// safe retry without server-side duplication.
  ///
  /// Backend stores driverDistance as canonical METERS — no conversion
  /// happens here; the UI layer handles display-unit conversion.
  Future<ProfileDTO> updateProfile(
    UpdateProfileRequest request, {
    String? idempotencyKey,
  }) async {
    final response = await _apiClient.put(
      '/profiles/me',
      body: request.toJson(),
      idempotencyKey: idempotencyKey,
    );
    return ProfileDTO.fromJson(response as Map<String, dynamic>);
  }
}

File: apps/api/src/main/java/vnpt/vsp/module/profile/ProfileService.java
package vnpt.vsp.module.profile;

import vnpt.vsp.module.profile.dto.GolferProfileResponse;
import vnpt.vsp.module.profile.dto.UpdateGolferProfileRequest;

/**
 * Profile module public service interface.
 * Exposes golfer profile and preference operations.
 * No module may directly @Autowired an Impl from another module — only this interface.
 */
public interface ProfileService {

    /**
     * Get the authenticated golfer's profile, auto-creating a default profile if none exists.
     * Per Story 2.3 AC-1: returns all profile fields.
     * Per Story 2.3 AC-2: returns canonical values (stored in meters) and display-unit preference.
     *
     * @param golferAccountId the authenticated golfer's account ID (from security principal)
     * @return the golfer's profile
     */
    GolferProfileResponse getProfile(Long golferAccountId);

    /**
     * Update the authenticated golfer's profile.
     * Per Story 2.3 AC-1: all golf preference fields are updatable.
     * Per Story 2.3 AC-2: canonical meters values are preserved; only display-unit preference changes.
     *
     * @param golferAccountId the authenticated golfer's account ID (from security principal)
     * @param request         the update request with fields to change
     * @return the updated profile
     */
    GolferProfileResponse updateProfile(Long golferAccountId, UpdateGolferProfileRequest request);
}
